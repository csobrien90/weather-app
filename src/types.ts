export interface WeatherData {
	generatedAt: string,
	periods: { 
		number: number,
		name: string,
		detailedForecast: string,
		temperature: number,
		windSpeed: string
	}[]
}

export interface WeatherProps extends WeatherData {
	locationData: LocationData
}

export interface WeatherDataExtended extends WeatherData {
	displayName: string
}

export interface LocationData {
	formattedAddress: string,
	displayName?: string
}


// {
//     "type": "Feature",
//     "properties": {
//         "country": "Brazil",
//         "country_code": "br",
//         "region": "Southeast Region",
//         "state": "São Paulo",
//         "county": "Região Metropolitana de Campinas",
//         "city": "Sumaré",
//         "municipality": "Região Imediata de Campinas",
//         "postcode": "13170-001",
//         "district": "Sumaré",
//         "datasource": {
//             "sourcename": "openstreetmap",
//             "attribution": "© OpenStreetMap contributors",
//             "license": "Open Database License",
//             "url": "https://www.openstreetmap.org/copyright"
//         },
//         "state_code": "SP",
//         "result_type": "postcode",
//         "lon": -47.2702083,
//         "lat": -22.8190147,
//         "parent_as_place_id": true,
//         "formatted": "Sumaré (Sumaré) - SP, 13170-001, Brazil",
//         "address_line1": "Sumaré (Sumaré) - SP",
//         "address_line2": "13170-001, Brazil",
//         "category": "administrative",
//         "timezone": {
//             "name": "America/Sao_Paulo",
//             "offset_STD": "-03:00",
//             "offset_STD_seconds": -10800,
//             "offset_DST": "-03:00",
//             "offset_DST_seconds": -10800
//         },
//         "plus_code": "589J5PJH+9W",
//         "plus_code_short": "JH+9W Sumaré, Região Metropolitana de Campinas, Brazil",
//         "rank": {
//             "confidence": 1,
//             "confidence_city_level": 1,
//             "match_type": "full_match"
//         },
//         "place_id": "51cbcd812f96a247c059797187f2aad136c0f00101f901dbbe560000000000c00207"
//     },
//     "geometry": {
//         "type": "Point",
//         "coordinates": [
//             -47.2702083,
//             -22.8190147
//         ]
//     },
//     "bbox": [
//         -47.3706769,
//         -22.9083958,
//         -47.2332661,
//         -22.7893005
//     ]
// }
export interface autocompleteLocationData {
	"properties": {
		"country": string,
		"country_code": string,
		"region": string,
		"state": string,
		"county": string,
		"city": string,
		"municipality": string,
		"postcode": string,
		"district": string,
		"datasource": {
			"sourcename": string,
			"attribution": string,
			"license": string,
			"url": string
		},
		"state_code": string,
		"result_type": string,
		"lon": number,
		"lat": number,
		"parent_as_place_id": boolean,
		"formatted": string,
		"address_line1": string,
		"address_line2": string,
		"category": string,
		"timezone": {
			"name": string,
			"offset_STD": string,
			"offset_STD_seconds": number,
			"offset_DST": string,
			"offset_DST_seconds": number
		},
		"plus_code": string,
		"plus_code_short": string,
		"rank": {
			"confidence": number,
			"confidence_city_level": number,
			"match_type": string
		},
		"place_id": string
	}
}