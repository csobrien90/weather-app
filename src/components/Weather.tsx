export default function Weather({
	generatedAt,
	periods,
	locationData
}: {
	generatedAt: string,
	periods: { 
		number: number,
		name: string,
		detailedForecast: string,
		temperature: number,
		windSpeed: string
	}[],
	locationData: {formattedAddress: string, displayName: string}
}) {
	if (!periods || !periods.length || !generatedAt) {
		return null;
	}

	// Format the generatedAt date
	const date = new Date(generatedAt);
	const options: Intl.DateTimeFormatOptions = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric', hour: 'numeric', minute: 'numeric' };
	const formattedDate = date.toLocaleTimeString(undefined, options);

	return (
		<section>
			<h2>Forecast for {locationData.displayName.split(',').splice(1).join(',').replace(', United States of America', '')}</h2>
			<p>Last updated on: {formattedDate}</p>
			{periods.map((period, index) => (
				<article key={index}>
					<h3>{period.name}</h3>
					<p>{period.detailedForecast}</p>
					<p>Temperature: {period.temperature}Â°F</p>
					<p>Wind Speed: {period.windSpeed}</p>
				</article>
			))}
		</section>
	)
}