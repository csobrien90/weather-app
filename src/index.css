@import url('https://fonts.googleapis.com/css2?family=Merriweather+Sans:ital,wght@0,300..800;1,300..800&family=Open+Sans:ital,wght@0,300;0,400;0,500;0,600;0,700;0,800;1,300;1,400;1,500;1,600;1,700;1,800&family=Varela+Round&display=swap');

:root {
	--gray: hsl(158, 7%, 52%)	;
	--orange: hsl(30, 45%, 50%);
	--dk-orange: hsl(30, 45%, 45%);
	--beige: hsl(52, 33%, 69%);
	--green: hsl(75, 37%, 53%);

	--primary: var(--green);
	--secondary: var(--beige);
	--tertiary: var(--orange);
	--neutral: var(--gray);
}

* {
	box-sizing: border-box;
	font-family: "Merriweather Sans", sans-serif;
	margin: 0;
	padding: 0;
}

html {
	scroll-behavior: smooth;
}

body {
	font-size: 10px;
	line-height: 1.5;
	padding: 1em 2em;
	max-width: 1200px;
	background-color: var(--secondary)
}

h1 {
	font-size: 5em;
}

h2 {
	font-size: 3em;
}

h3 {
	font-size: 2.5em;
}

h4 {
	font-size: 2em;
}

h5 {
	font-size: 1.6em;
}

p, input, label, button {
	font-size: 1.6em;
	display: block;
}

main, article, section, form {
	display: grid;
	gap: 1em;
	padding: 1em 0;
	justify-content: flex-start;
}

input {
	padding: .5em;
	border: 1px solid var(--neutral);
}

button {
	background-color: var(--tertiary);
	border: none;
	padding: 1em 2em;
	cursor: pointer;
}

button:hover, button:active {
	color: white;
	background-color: var(--dk-orange);
	box-shadow: inset 0 0 10px -2px var(--orange);
}

@media screen and (min-width: 768px){
	body {
		font-size: 12px;
	}
}

@media screen and (min-width: 1024px){
	body {
		font-size: 15px;
	}
}