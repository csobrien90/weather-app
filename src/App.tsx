import { useState } from 'react';

import LocationInput from './components/LocationInput'
import Weather from './components/Weather'

import './App.css'

export default function App() {
	const [locationData, setLocationData] = useState({formattedAddress: '', displayName: ''});

	const [weatherData, setWeatherData]: [any, any] = useState({
		generatedAt: '',
		periods: []
	});

	return (
		<main>
			<h1>Weather App</h1>
			<LocationInput locationData={locationData} setLocationData={setLocationData} setWeatherData={setWeatherData} />
			<Weather locationData={locationData} {...weatherData} />
		</main>
	)
}