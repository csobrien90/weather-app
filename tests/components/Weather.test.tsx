import React from 'react';
import { expect, test } from 'vitest';
import renderer from 'react-test-renderer'
import Weather from '../../src/components/Weather';
import { WeatherDays } from '../../src/types';

const sampleDay: WeatherDays = {
	'Today': {
		day: {
			number: 1,
			name: 'Today',
			icon: 'sunny',
			startTime: '2021-08-01T06:00:00-05:00',
			endTime: '2021-08-01T18:00:00-05:00',
			detailedForecast: 'Sunny',
			temperature: 80,
			windSpeed: '5 mph',
			windDirection: 'N',
			temperatureUnit: 'F',
		}
	}
}

test('Weather renders correctly', () => {
	const component = renderer.create(
		<Weather
			generatedAt='2021-08-01T12:00:00-05:00'
			days={sampleDay}
			locationData={{
				formattedAddress: '123 Main St, Springfield, IL 62701',
				displayName: '123 Main St, Springfield, IL'
			}}
		/>
	).toJSON();

	expect(component).to.have.property('type', 'section');
	expect(component).to.matchSnapshot();
})

test('Weather does not render if there are no days', () => {
	const component = renderer.create(
		<Weather
			generatedAt='2021-08-01T12:00:00-05:00'
			days={{}}
			locationData={{
				formattedAddress: '123 Main St, Springfield, IL 62701',
				displayName: '123 Main St, Springfield, IL'
			}}
		/>
	).toJSON();

	expect(component).to.be.null;
})

test('Weather does not render if there is no generatedAt', () => {
	const component = renderer.create(
		<Weather
			generatedAt=''
			days={sampleDay}
			locationData={{
				formattedAddress: '123 Main St, Springfield, IL 62701',
				displayName: '123 Main St, Springfield, IL'
			}}
		/>
	).toJSON();

	expect(component).to.be.null;
})

test('Weather does not render if there is no locationData', () => {
	const component = renderer.create(
		<Weather
			generatedAt='2021-08-01T12:00:00-05:00'
			days={sampleDay}

			// @ts-ignore - Testing the component's behavior when locationData is not provided
			locationData={{}}
		/>
	).toJSON();

	expect(component).to.be.null;
})