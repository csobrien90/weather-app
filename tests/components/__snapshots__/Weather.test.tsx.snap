// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Weather renders correctly 1`] = `
<section>
  <h2>
    Forecast for 
    123 Main St, Springfield
  </h2>
  <p>
    Last updated on: 
    Sunday, August 1, 2021 at 1:00 PM
  </p>
  <article
    className="weather-card"
  >
    <img
      alt="sunny"
      src="/src/assets/images/sun.svg"
    />
    <div>
      <h3>
        Today
      </h3>
      <p>
        <span>
          <img
            alt="sun"
            src="/src/assets/images/day.svg"
          />
           
          80
          °
          F
        </span>
      </p>
      <span
        className="description"
      >
        <p>
          Sunny
        </p>
      </span>
    </div>
  </article>
</section>
`;
